subsys/bluetooth/services/ots/ots_dir_list.c

```diff
@@ -220,26 +191,72 @@ void bt_ots_dir_list_init(struct bt_ots_dir_list **dir_list, void *obj_manager)
 
 	memset(&dir_list_obj->metadata, 0, sizeof(dir_list_obj->metadata));
 	dir_list_obj->metadata.name = dir_list_obj_name;
-	dir_list_obj->metadata.size.alloc = sizeof((*dir_list)->_content);
+	dir_list_obj->metadata.size.alloc = DIR_LIST_MAX_SIZE;
 	dir_list_obj->metadata.type.uuid.type = BT_UUID_TYPE_16;
 	dir_list_obj->metadata.type.uuid_16.val = BT_UUID_OTS_DIRECTORY_LISTING_VAL;
 	BT_OTS_OBJ_SET_PROP_READ(dir_list_obj->metadata.props);
 
 	(*dir_list)->dir_list_obj = dir_list_obj;
 
-	/* Set size in dir_list_encode */
-	dir_list_encode(*dir_list, obj_manager);
+	dir_list_update_size(*dir_list, obj_manager);
 }
 
-ssize_t bt_ots_dir_list_content_get(struct bt_ots_dir_list *dir_list, void **data,
-				size_t len, off_t offset)
+ssize_t bt_ots_dir_list_content_get(struct bt_ots_dir_list *dir_list, void *obj_manager,
+				    void **data, size_t len, off_t offset)
 {
-	if (offset >= dir_list->net_buf.len) {
-		*data = NULL;
-		return 0;
-	}
+	struct bt_gatt_ots_object *obj;
+	int err;
+	size_t buffer_offset = 0;
+	size_t rec_len = 0;
+	size_t room = 0;
+	bool found = false;
+
+	err = bt_gatt_ots_obj_manager_first_obj_get(obj_manager, &obj);
+
+	__ASSERT(err == 0 && obj == dir_list->dir_list_obj,
+		 "Expecting first object to be the Directory Listing Object");
+
+	do {
+		/* keep resetting the buffer until we find the first object record that
+		 * is of intereset.
+		 */
+		if (!found) {
+			buffer_offset += rec_len;
+
+			net_buf_simple_init_with_data(&dir_list->net_buf,
+						      dir_list->_content,
+						      sizeof(dir_list->_content));
+			net_buf_simple_reset(&dir_list->net_buf);
+			room = sizeof(dir_list->_content);
+		}
+
+		rec_len = dir_list_object_encode(obj, &dir_list->net_buf);
 
-	*data = dir_list->net_buf.data + offset;
+		/* if this record contains at lease one byte that the client is asking
+		 * for then we have found our first entry and we should stop resetting
+		 * the buffer. We could only keep the bytes that are off intereset to
+		 * the client but then we would have a buffer that isn't a multiple of
+		 * object records and that would make things more complicated.
+		 */
+		found = buffer_offset + rec_len > offset;
```

Comment:
Do we encode all objects until we find the right one? 

Let's say we have 100 objects, and we have already sent the directory listing of the first 80. 
This looks like it will, when `bt_ots_dir_list_content_get` is called, encode all of the 80 again before finally reaching the 81st object and finally getting to "new data". Is this how this is working, and intended to work?

Reply:
When we have 10 objects, then `bt_ots_dir_list_content_get` should be called at most 10 times (depending on the MTU), and likewise `dir_list_object_encode` should be called a most 10 times.

Reply:
> Do we encode all objects until we find the right one?
> 
> Let's say we have 100 objects, and we have already sent the directory listing of the first 80.
This looks like it will, when bt_ots_dir_list_content_get is called, encode all of the 80 again before finally reaching the 81st object and finally getting to "new data". Is this how this is working, and intended to work?

That is correct, that is how it was intended to work.

> When we have 10 objects, then bt_ots_dir_list_content_get should be called at most 10 times (depending on the MTU), and likewise dir_list_object_encode should be called a most 10 times.

This would require us to maintain state across different L2CAP reads. Also as the client is not obligated to read the object sequentially we would need to have logic to detect that the read is not sequential and have to find our read location again. I had opted to keep this simple as this was intended to be a memory efficient implementation.

We can add logic to keep the current object record we are in and what offset that corresponds to, then we can search forward or backwards depending on the next offset requested. We would have to add some heuristics such as reset to first object on a directory listing select.

I can look at adding this as part of this PR if you think the performance hit is unacceptable or will follow up with a separate PR if the current performance is acceptable.

Reply:
I think we need to keep some state for this to be useful, as this scales very poorly. 

In the case that we serve e.g. 1000 objects (perhaps not that uncommon if running on a media server), and we assume worst case (i.e. each metadata is 172 octets), then when reading the final object metadata (which should just be 172 octets), then we are in fact first going through 999 other objects and copying them (as part of dir_list_object_encode), which responds to copying 172k octets instead of just 172. 

Ideally we would at total only encode 172k octets for the entire directory listing, but with this approach we will do a grand total copy of 86086k octets (i.e. 86MiB instead of 172KiB!!!).  

I reckon that if you tried a sample using this implementation with 1000 auto-generated objects, you will also see a significant increase in time. 

Using a state and the offset should provide us with means to improve this significantly, without using much more memory. Alternatively, at the very least, when traversing through the objects we should not encode each object as we go along, but only those starting from the "object of interest". 

Before we start working too much on, I'd like you to run a few tests on some decent size data sets (e.g. 100 or maybe even 1000 objects) and then see what kind of time/performance you are seeing. If it isn't too bad, then we can move along with this as is, and postpone the optimizations.

Reply:
Sounds good! I will collect some performance data, I doubt I can get to 1000 given that I am using an nrf52dk development board and it doesn't have enough memory for 1000 objects.